websocketserver.go

```diff
@@ -23,48 +26,109 @@ type protocol struct {
 	// Serializer for binary frames.  Defaults to MessagePackSerializer.
 	BinarySerializer serialize.Serializer
 
-	protocols map[string]protocol
+	router Router
 
-	log stdlog.StdLog
+	protocols map[string]protocol
+	log       stdlog.StdLog
 }
 
 // NewWebsocketServer takes a router instance and creates a new websocket
-// server.
+// server.  To run the websocket server, call one of the server's
+// ListenAndServe methods:
+//
+//     s := NewWebsocketServer(r)
+//     closer, err := s.ListenAndServe(address)
+//
+// Or, use the various ListenAndServe functions provided by net/http.  This
+// works because WebsocketServer implements the http.Handler interface:
+//
+//     s := NewWebsocketServer(r)
+//     server := &http.Server{
+//         Handler: s,
+//         Addr:    address,
+//     }
+//     server.ListenAndServe()
 func NewWebsocketServer(r Router) *WebsocketServer {
 	s := &WebsocketServer{
-		Router:    r,
+		router:    r,
 		protocols: map[string]protocol{},
 		log:       r.Logger(),
 	}
-
 	s.Upgrader = &websocket.Upgrader{}
-	s.AddProtocol(jsonWebsocketProtocol, websocket.TextMessage,
+	s.addProtocol(jsonWebsocketProtocol, websocket.TextMessage,
 		&serialize.JSONSerializer{})
-	s.AddProtocol(msgpackWebsocketProtocol, websocket.BinaryMessage,
+	s.addProtocol(msgpackWebsocketProtocol, websocket.BinaryMessage,
 		&serialize.MessagePackSerializer{})
+
 	return s
 }
 
-// AddProtocol registers a serializer for protocol and payload type.
-func (s *WebsocketServer) AddProtocol(proto string, payloadType int, serializer serialize.Serializer) error {
-	s.log.Println("AddProtocol:", proto)
-	if payloadType != websocket.TextMessage && payloadType != websocket.BinaryMessage {
-		return fmt.Errorf("invalid payload type: %d", payloadType)
+// ListenAndServe listens on the specified TCP address and starts a goroutine
+// that accept new client connections until the returned io.closer is closed.
+func (s *WebsocketServer) ListenAndServe(address string) (io.Closer, error) {
+	l, err := net.Listen("tcp", address)
+	if err != nil {
+		s.log.Print(err)
+		return nil, err
 	}
-	if _, ok := s.protocols[proto]; ok {
-		return errors.New("protocol already registered: " + proto)
+	addr := l.(*net.TCPListener).Addr()
+
+	// Run service on configured port.
+	server := &http.Server{
+		Handler: s,
+		Addr:    addr.String(),
 	}
-	s.protocols[proto] = protocol{payloadType, serializer}
-	s.Upgrader.Subprotocols = append(s.Upgrader.Subprotocols, proto)
-	return nil
+	go server.Serve(l)
+	return l, nil
+}
+
+// ListenAndServeTLS listens on the specified TCP address and starts a
+// goroutine that accept new TLS client connections until the returned
+// io.closer is closed.
+func (s *WebsocketServer) ListenAndServeTLS(address string, tlsConfig *tls.Config, certFile, keyFile string) (io.Closer, error) {
+	l, err := net.Listen("tcp", address)
+	if err != nil {
+		s.log.Print(err)
+		return nil, err
+	}
+	addr := l.(*net.TCPListener).Addr()
```

Comment:
variable not needed. set `l.Addr()` directly on `http.Server.Addr`

Reply:
done